src/storage/conversion/DynamicJsonToTemplateConverter.ts

```diff
@@ -0,0 +1,99 @@
+import type { Term, NamedNode } from 'rdf-js';
+import { BasicRepresentation } from '../../ldp/representation/BasicRepresentation';
+import type { Representation } from '../../ldp/representation/Representation';
+import { RepresentationMetadata } from '../../ldp/representation/RepresentationMetadata';
+import type { ValuePreferences } from '../../ldp/representation/RepresentationPreferences';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { APPLICATION_JSON } from '../../util/ContentTypes';
+import { NotImplementedHttpError } from '../../util/errors/NotImplementedHttpError';
+import { readableToString } from '../../util/StreamUtil';
+import type { TemplateEngine } from '../../util/templates/TemplateEngine';
+import { CONTENT_TYPE, CONTENT_TYPE_TERM, SOLID_META } from '../../util/Vocabularies';
+import { getConversionTarget } from './ConversionUtil';
+import { RepresentationConverter } from './RepresentationConverter';
+import type { RepresentationConverterArgs } from './RepresentationConverter';
+
+/**
+ * Converts JSON data by using it as input parameters for rendering a template.
+ * The `extension` field can be used to only support a specific type of templates,
+ * such as ".ejs" for EJS templates.
+ *
+ * To find the templates it expects the Representation metadata to contain `SOLID_META.template` triples,
+ * with the objects being the template paths.
+ * For each of those templates there also needs to be a CONTENT_TYPE triple
+ * describing the content-type of that template.
+ *
+ * The output of the result depends on the content-type matched with the template.
+ */
+export class DynamicJsonToTemplateConverter extends RepresentationConverter {
+  private readonly templateEngine: TemplateEngine;
+  private readonly extension: string;
+
+  public constructor(templateEngine: TemplateEngine, extension = '') {
```

Comment:
I find the extension weird. Is this a proxy for mime types?

Reply:
Sort of I guess. The idea was that I wanted to prevent the template engine being called on templates it didn't support (so only use the engine on `.ejs` templates if it is an EJS engine). But thinking about it again now, this should not be here. Instead, this is a very good case on why we should make the template engines AsyncHandlers. Then the engine itself can check the file extension in its canHandle function so the EJS engine only supports `.ejs` templates. We could then even have a waterfall handler with both EJS and Handlebars engine so we have a single engine that supports both.